File: tests/native/test_event.py
# Copyright 2026 RynnRCP Authors. All rights reserved.
# Tests for rynnrcp.native.event

import sys
import os
import threading
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from rynnrcp.native.event import Event, WaitResult


def test_create_close():
    ev = Event()
    assert ev.is_valid()
    ev.close()
    assert not ev.is_valid()
    print("[PASS] test_create_close")


def test_signal_wait():
    ev = Event()
    assert ev.is_valid()

    ok = ev.signal()
    assert ok

    result = ev.wait(timeout_ms=1000)
    assert result == WaitResult.SIGNALED

    ev.close()
    print("[PASS] test_signal_wait")


def test_timeout():
    ev = Event()
    assert ev.is_valid()

    start = time.monotonic()
    result = ev.wait(timeout_ms=50)
    elapsed_ms = (time.monotonic() - start) * 1000

    assert result == WaitResult.TIMEOUT
    assert elapsed_ms >= 30, f"Timeout too fast: {elapsed_ms:.1f}ms"

    ev.close()
    print(f"[PASS] test_timeout (waited {elapsed_ms:.0f}ms)")


def test_nonblocking_poll():
    ev = Event()
    assert ev.is_valid()

    # No signal -> timeout
    result = ev.wait(timeout_ms=0)
    assert result == WaitResult.TIMEOUT

    # Signal then poll -> signaled
    ev.signal()
    result = ev.wait(timeout_ms=0)
    assert result == WaitResult.SIGNALED

    ev.close()
    print("[PASS] test_nonblocking_poll")


def test_cross_thread_signal():
    ev = Event()
    received = [False]

    def waiter():
        result = ev.wait(timeout_ms=5000)
        if result == WaitResult.SIGNALED:
            received[0] = True

    t = threading.Thread(target=waiter)
    t.start()

    # Give waiter time to enter wait
    time.sleep(0.05)
    ok = ev.signal()
    assert ok

    t.join(timeout=5)
    assert received[0], "Waiter did not receive signal"

    ev.close()
    print("[PASS] test_cross_thread_signal")


def test_multiple_signals():
    ev = Event()
    for _ in range(5):
        ev.signal()

    result = ev.wait(timeout_ms=100)
    assert result == WaitResult.SIGNALED

    ev.close()
    print("[PASS] test_multiple_signals")


def test_closed_event():
    ev = Event()
    ev.close()

    assert not ev.signal()
    result = ev.wait(timeout_ms=0)
    assert result == WaitResult.ERROR
    print("[PASS] test_closed_event")


def test_auto_reset():
    """After wait consumes the signal, subsequent wait should timeout."""
    ev = Event()
    ev.signal()

    result = ev.wait(timeout_ms=100)
    assert result == WaitResult.SIGNALED

    # Signal was consumed, next wait should timeout
    result = ev.wait(timeout_ms=50)
    assert result == WaitResult.TIMEOUT

    ev.close()
    print("[PASS] test_auto_reset")


if __name__ == "__main__":
    print("=== event tests ===")
    test_create_close()
    test_signal_wait()
    test_timeout()
    test_nonblocking_poll()
    test_cross_thread_signal()
    test_multiple_signals()
    test_closed_event()
    test_auto_reset()
    print("=== All event tests passed ===")

File: tests/cpp/test_event.cc
// Copyright 2026 RynnRCP Authors. All rights reserved.
// Tests for rynnrcp/native/cpp/event.h

#include "event.h"

#include <cassert>
#include <cstdio>
#include <thread>
#include <chrono>

using namespace rynnrcp::native;

void test_create_close() {
    EventHandle ev = EventCreate();
    assert(ev != kInvalidEvent);
    EventClose(ev);
    assert(ev == kInvalidEvent);
    printf("[PASS] test_create_close\n");
}

void test_signal_wait() {
    EventHandle ev = EventCreate();
    assert(ev != kInvalidEvent);

    // Signal then wait should succeed immediately
    bool ok = EventSignal(ev);
    assert(ok);

    WaitResult result = EventWait(ev, 1000);  // 1 second timeout
    assert(result == WaitResult::kSignaled);

    EventClose(ev);
    printf("[PASS] test_signal_wait\n");
}

void test_timeout() {
    EventHandle ev = EventCreate();
    assert(ev != kInvalidEvent);

    // Wait without signal should timeout
    auto start = std::chrono::steady_clock::now();
    WaitResult result = EventWait(ev, 50);  // 50ms timeout
    auto elapsed = std::chrono::steady_clock::now() - start;

    assert(result == WaitResult::kTimeout);
    // Should have waited at least ~40ms (allow some tolerance)
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();
    assert(ms >= 30);

    EventClose(ev);
    printf("[PASS] test_timeout (waited %lld ms)\n", static_cast<long long>(ms));
}

void test_nonblocking_poll() {
    EventHandle ev = EventCreate();
    assert(ev != kInvalidEvent);

    // Non-blocking poll (timeout=0) with no signal -> timeout
    WaitResult result = EventWait(ev, 0);
    assert(result == WaitResult::kTimeout);

    // Signal then non-blocking poll -> signaled
    EventSignal(ev);
    result = EventWait(ev, 0);
    assert(result == WaitResult::kSignaled);

    EventClose(ev);
    printf("[PASS] test_nonblocking_poll\n");
}

void test_cross_thread_signal() {
    EventHandle ev = EventCreate();
    assert(ev != kInvalidEvent);

    bool received = false;

    std::thread waiter([&]() {
        WaitResult result = EventWait(ev, 5000);  // 5 sec max
        if (result == WaitResult::kSignaled) {
            received = true;
        }
    });

    // Give waiter time to enter wait
    std::this_thread::sleep_for(std::chrono::milliseconds(50));

    bool ok = EventSignal(ev);
    assert(ok);

    waiter.join();
    assert(received);

    EventClose(ev);
    printf("[PASS] test_cross_thread_signal\n");
}

void test_multiple_signals() {
    EventHandle ev = EventCreate();
    assert(ev != kInvalidEvent);

    // Multiple signals before wait
    for (int i = 0; i < 5; ++i) {
        EventSignal(ev);
    }

    // Should be able to consume at least one signal
    WaitResult result = EventWait(ev, 100);
    assert(result == WaitResult::kSignaled);

    EventClose(ev);
    printf("[PASS] test_multiple_signals\n");
}

int main() {
    printf("=== event tests ===\n");
    test_create_close();
    test_signal_wait();
    test_timeout();
    test_nonblocking_poll();
    test_cross_thread_signal();
    test_multiple_signals();
    printf("=== All event tests passed ===\n");
    return 0;
}
